File: java/sort/linked_list/QuickSort.java
package sort.linked_list;

public class QuickSort {
    public void sort(ListNode head) {
        sort(head, head.next, null);
    }

    private void sort(ListNode head, ListNode start, ListNode end) {
        if (start != end && start.next != end) {
            ListNode partition = getPartition(head, start, end);
            sort(head, head.next, partition);
            sort(partition, partition.next, end);
        }
    }

    private ListNode getPartition(ListNode head, ListNode start, ListNode end) {
        ListNode headLeft = new ListNode(0);
        ListNode headRight = new ListNode(0);

        ListNode left = headLeft;
        ListNode right = headRight;
        int pivot = start.value;
        for (ListNode node = start.next; node != end; node = node.next) {
            if (node.value < pivot) {
                left.next = node;
                left = node;
            } else {
                right.next = node;
                right = node;
            }
        }
        left.next = start;
        right.next = end;

        head.next = headLeft.next;
        start.next = headRight.next;
        return start;
    }

    public class ListNode {
        public int value = 0;
        public ListNode next = null;

        public ListNode(int value) {
            this.value = value;
        }
    }
}

File: py/sort/linked_list/QuickSort.py
#!/usr/bin/env python
# -*- coding:gbk -*-

class ListNode:
    def __init__(self, value: int):
        self.value, self.next = value, None

class QuickSort:
    def sort(self, head: ListNode) -> None:
        self.__sort(head, head.next, None)

    def __sort(self, head: ListNode, start: ListNode, end: ListNode) -> None:
        if start == end or start.next == end:
            return

        partition = self.__getPartition(head, start, end)
        self.__sort(head, head.next, partition)
        self.__sort(partition, partition.next, end)

    def __getPartition(self, head: ListNode, start: ListNode, end: ListNode) -> None:
        headLeft, headRight = ListNode(0), ListNode(0)
        pivot, left, right, node = start.value, headLeft, headRight, start.next
        while node != end:
            if node.value < pivot:
                left.next, left = node, node
            else:
                right.next, right = node, node
            node = node.next
        left.next, right.next = start, end

        head.next, start.next = headLeft.next, headRight.next
        return start
